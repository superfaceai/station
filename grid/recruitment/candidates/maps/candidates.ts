/* eslint-disable jest/no-export */

import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';
import { readFileSync } from 'fs';

const sampleCandidate = {
  name: 'John Doe',
  firstName: 'John',
  lastName: 'Doe',

  email: 'john.doe@fakemail.com',
  phone: '1-859-557-6573',
  address: '25772 Gustave Shore, Iowa, USA',

  education: [
    {
      degree: 'MBA',
      school: 'University of Pennsylvania',
      fieldOfStudy: null,
      startedAt: '2008-03-01',
      endedAt: '2011-03-30',
    },
    {
      degree: 'B.S.',
      school: 'University of Chicago',
      fieldOfStudy: 'Marketing Communication & Economics',
      startedAt: '2004-09-01',
      endedAt: '2007-03-30',
    },
  ],

  workExperience: [
    {
      position: 'Sales Director',
      summary: 'Summary of a work experience at Test Company',
      company: 'Test Company',
      industry: 'Telecommunications',
      current: false,
      startedAt: '2011-03-01',
      endedAt: '2014-03-30',
    },
  ],

  cv: {
    name: 'cv-sample.pdf',
    data: readFileSync('./grid/recruitment/candidates/cv-sample.pdf', {
      encoding: 'base64',
    }),
  },

  links: [
    {
      name: 'twitter',
      url: 'https://twitter.com/candidate.username',
    },
    {
      name: 'Portfolio',
      url: 'https://url.to.portfolio',
    },
  ],
};

export const candidatesTest = (
  provider: string,
  jobIds: { valid: string; invalid: string },
  options?: RecordingProcessOptions
): void => {
  describe(`recruitment/candidates/${provider}`, () => {
    let superface: SuperfaceTest;

    describe('CreateCandidate', () => {
      beforeAll(() => {
        superface = new SuperfaceTest({
          profile: 'recruitment/candidates',
          provider,
          useCase: 'CreateCandidate',
          testInstance: expect,
        });
      });

      describe('when specified job does exist', () => {
        it('performs correctly', async () => {
          const page1 = await superface.run(
            {
              input: {
                jobId: jobIds.valid,
                ...sampleCandidate,
              },
              testName: 'page 1',
            },
            options
          );

          expect(page1.isOk).toBeTruthy();
          expect(page1).toMatchSnapshot();
        });
      });

      describe('when specified job does not exist', () => {
        it('returns error', async () => {
          await expect(
            superface.run(
              {
                input: {
                  jobId: jobIds.invalid,
                  firstName: 'Demo',
                  lastName: 'Testing',
                  email: 'demo_testing@fakemail.com',
                },
              },
              options
            )
          ).resolves.toMatchSnapshot();
        });
      });
    });
  });
};
