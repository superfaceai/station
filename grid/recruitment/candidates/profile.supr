name = "recruitment/candidates"
version = "1.0.0"

"""
Create Candidate
Creates an application for specific job. The candidate may (depends on the provider implementation) receive notification that her/his application was successfully accepted by the company.
"""
usecase CreateCandidate {

  input {
    """
    Job Id
    Id of a job, to which candidate applies to.
    """
    jobId! string!

    """
    First Name
    First name of a candidate.
    """
    firstName! string!

    """
    Last Name
    Last name of a candidate.
    """
    lastName! string!

    """
    Email
    Email of a candidate.
    """
    email! string!

    """
    Phone
    Phone number of a candidate.
    """
    phone string

    """
    Address
    Address of a candidate.
    """
    address string	

    """
    Education
    Array of education experiences.
    """
    education [Education]

    """
    Work Experience
    Array of work experiences.
    """
    workExperience [WorkExperience]

    """
    CV (Resume)
    Object containing either url to a CV or data of a CV.
    """
    cv { url string } | { name string, data string }

    """
    Links
    Array of links to different profiles or portfolios of a candidate.
    """
    links [Link]

    """
    Answers
    Array of answers to questions connected to a job that candidate applies to.
    """
    answers [Answer]

    """
    Stage Id
    Unique identifier of a stage, in which candidate is in.
    """
    stageId string
  }

  result {
    """
    Id
    Unique identifier of newly created candidate.
    """
    id! string!

    """
    Job Id
    Unique identifier of a job, newly created candidate applied for.
    """
    jobId! string!
  }!

  error Error
}

"""
Education
Object representing educational experience of a candidate.
"""
model Education {
  id string
  school string
  degree string
  fieldOfStudy string
  
  startedAt string
  endedAt string
}

"""
Work Experience
Object representing work experience of a candidate.
"""
model WorkExperience {
  id string
  position string
  summary string
  company string
  industry string
  current boolean

  startedAt string
  endedAt string
}

"""
Answer
Object representing a answer to specific question from a job.
"""
model Answer {
  questionId string
  type enum {
    text
    textarea,
    number
    boolean
    multiple_choice
    multiple_select
    date
    file
  }
  value
}

"""
Link
Object representing a link specified by candidate.
"""
model Link {
  name string
  url string
}

model Candidate {
  """
  Id
  Unique identifier of a candidate
  """
  id! string!
  
  """
  Job Id
  Unique identifier of a job, candidate applied for.
  """
  jobId! string!

  """
  Job Title
  Title of a job, candidate applied for.
  """
  jobTitle string

  """
  First Name
  First name of a candidate.
  """
  firstName! string!

  """
  Last Name
  Last name of a candidate.
  """
  lastName! string!

  """
  Full Name
  Full name of a candidate.
  """
  fullName string

  """
  Email
  Email of a candidate.
  """
  email! string!

  """
  Phone
  Phone number of a candidate.
  """
  phone string

  """
  Address
  Address of a candidate.
  """
  address string	

  """
  Links
  Array of links specified by a candidate. Can contain links to social profiles, portfolios or other.
  """
  links [Link]

  """
  Experiences
  Array of work experiences specified by a candidate.
  """
  experiences [WorkExperience]
  
  """
  Education
  Array of education experiences specified by a candidate.
  """
  education [Education]

  """
  CV (Resume)
  Object containing either url to a CV or data of a CV.
  """
  cv { url string } | { name string, data string }

  """
  CV Url
  Url pointing to CV of a candidate.
  """
  cvUrl string

  """
  Answers
  Array of answers to questions connected to a job that candidate applies to.
  """
  answers [Answer]

  """
  Stage
  Name of a stage, candidate is currently in.
  """
  stage string

  """
  Stage Changes
  Array of stages, candidate once was in (except for current one).
  """
  stageChanges [string]

  """
  Created At
  ISO8601 timestamp when was a candidate created.
  """
  createdAt string
  
  """
  Updated At
  ISO8601 timestamp when was a candidate lastly updated.
  """
  updatedAt string
}

model Error {
  """
  Error Title
  A short, human-readable summary of the problem type.
  """
  title!
  """
  Error Detail
  A human-readable explanation specific to this occurrence of the problem.
  """
  detail
  """
  Code
  Error code.
  """
  code! ErrorCode!
  """
  Rate limit
  Rate limit details.
  """
  rateLimit RateLimit
}

model RateLimit {
  """
  Bucket
  Different parts of API may have different rate limits.
  Bucket identifies to which part of API the rate limits apply.
  """
  bucket string
  """
  Total requests
  Total requests available in the time window.
  """
  totalRequests number
  """
  Remaining requests
  Remaining requests available in the time window.
  """
  remainingRequests number
  """
  Remaining requests available in percents
  Remaining requests available in the time window in percents.
  """
  remainingRequestsPercentage number
  """
  Reset timestamp
  Timestamp when the rate limits will reset (in  Unix Timestamp format). Available only for providers without rolling rate limit window.
  """
  resetTimestamp number
}

model ErrorCode enum {
  """
  Bad request
  The request was invalid or cannot be otherwise served.
  """
  BadRequest
  """
  Not found
  The profileId does not exist.
  """
  NotFound
  """
  Rate limit reached
  Returned when a request cannot be served due to the user or app reached max requests quota.
  """
  RateLimitReached
  """
  Unauthenticated
  There was a problem authenticating your request. Check that you are passing valid credentials.
  """
  Unauthenticated
  """
  Unauthorized
  The request is understood, but it has been refused or access is not allowed.
  """
  Unauthorized
  """
  Unknown error
  Returned when an unexpected error occured.
  """
  UnknownError
}
