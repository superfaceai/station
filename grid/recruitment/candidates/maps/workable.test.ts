import { candidatesTest } from './candidates';

candidatesTest('workable', { valid: '56D021066C', invalid: 'NOT-EXISTING' });
