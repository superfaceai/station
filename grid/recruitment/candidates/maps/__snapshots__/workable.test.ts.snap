// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recruitment/candidates/workable CreateCandidate when specified job does exist performs correctly 1`] = `
Ok {
  "value": Object {
    "id": "d20b37b",
    "jobId": "56D021066C",
  },
}
`;

exports[`recruitment/candidates/workable CreateCandidate when specified job does not exist returns error 1`] = `
Err {
  "error": "MappedHTTPError: Expected HTTP error
Properties: {
  \\"title\\": \\"Not found\\",
  \\"detail\\": \\"Not found\\",
  \\"code\\": \\"NotFound\\",
  \\"rateLimit\\": {
    \\"bucket\\": \\"workable\\"
  }
}
AST Path: definitions[0].statements[0].responseHandlers[1].statements[1]
Original Map Location: Line 88, column 7",
}
`;
